hibernate-core/src/main/java/org/hibernate/query/sqm/tuple/internal/AnonymousTupleType.java

```diff
@@ -55,26 +56,51 @@ public class AnonymousTupleType<T>
 	private final String[] componentNames;
 	private final Map<String, Integer> componentIndexMap;
 
-	public AnonymousTupleType(SqmSubQuery<T> subQuery) {
-		this( extractSqmExpressibles( subQuery ) );
-	}
+	public AnonymousTupleType(SqmSelectQuery<T> selectQuery) {
+		final SqmSelectClause selectClause = selectQuery.getQueryPart()
+				.getFirstQuerySpec()
+				.getSelectClause();
 
-	public AnonymousTupleType(SqmSelectableNode<?>[] components) {
-		expressibles = new SqmBindableType<?>[components.length];
-		componentSourcePaths = new NavigablePath[components.length];
-		for ( int i = 0; i < components.length; i++ ) {
-			expressibles[i] = components[i].getNodeType();
-			if ( components[i] instanceof SqmPath<?> path ) {
-				componentSourcePaths[i] = path.getNavigablePath();
+		if ( selectClause == null || selectClause.getSelections().isEmpty() ) {
+			throw new IllegalArgumentException( "selectQuery has no selection items" );
+		}
+		// todo: right now, we "snapshot" the state of the selectQuery when creating this type, but maybe we shouldn't?
+		//  i.e. what if the selectQuery changes later on? Or should we somehow mark the selectQuery to signal,
+		//  that changes to the select clause are invalid after a certain point?
+
+		final List<SqmSelection<?>> selections = selectClause.getSelections();
+		final List<SqmSelectableNode<?>> selectableNodes = new ArrayList<>();
+		final List<String> aliases = new ArrayList<>();
+		if ( selections != null ) {
+			for ( SqmSelection<?> selection : selections ) {
+				final boolean compound = selection.getSelectableNode().isCompoundSelection();
+				selection.getSelectableNode().visitSubSelectableNodes( node -> {
+					selectableNodes.add( node );
+					if ( compound ) {
+						aliases.add( node.getAlias() );
+					}
+				} );
+				if ( !compound ) {
+					// for compound selections we use the sub-selectable nodes aliases
+					aliases.add( selection.getAlias() );
+				}
 			}
 		}
+
+		final SqmSelectableNode<?>[] components = selectableNodes.toArray(new SqmSelectableNode[0]);
```

Comment:
No need to transform this list to an array, just use `selectableNodes.get( i )`.

Reply:
Do you mean to use `selectableComponents` (list) instead of `components` (array) ? That's not a problem, but in this case `getTypeDescriptors` should be changed as well. Guess that this is OK?

Reply:
Ah sure, as I said this is a nitpick so it's fine to leave as-is. We could also change the method, or simply rely on the version accepting `SqmExpressible<?>[]` and pass the `expressibles` array we create just below, up to you.